include/onnxruntime/core/platform/EigenNonBlockingThreadPool.h

```diff
@@ -902,124 +842,265 @@ void EndParallelSectionInternal(PerThread &pt,
+// the worker thread that actually claims an item (e.g., if an item
+// initially assigned to thread T1 is stolen and executed by T2,
+// then T2 is assigned at the new preferred worker).
+//
+// Note that the hints are held in the _main_ thread that submits
+// work to the pool.  We assume that a thread is primarily
+// submitting work to just one pool, but allow for the pool to
+// change over time.  Hence we allow the hints vector to grow over
+// time.
+//
+//
+// A note on terminology used in the variable names here:
+//
+// dop - degree of parallelism, as seen by the user.  For instance
+//       dop=4 means 4 threads in total: 1 main thread that enters the
+//       loop, plus 1 dispatcher thread, plus 2 additional worker
+//       threads.
+//
+// par_idx - a thread's index within the loop, in the range [0,dop).
+//
+// num_threads_ - the number of worker threads in the thread pool.  A
+//       loop with dop=4 will be common on a pool with 3 threads
+//       (given that the main thread will also participate).
+//
+// q_idx - a worker queue index, in the range [0,num_threads_).
+ 
+void InitializePreferredWorkers(std::vector<int> &preferred_workers) {
+  static std::atomic<unsigned> next_worker;
+  // preferred_workers maps from a par_idx to a q_idx, hence we
+  // initialize slots in the range [0,num_threads_]
+  while (preferred_workers.size() <= num_threads_) {
+    preferred_workers.push_back(next_worker++ % num_threads_);
+  }
+}
+
+// Update the preferred worker for par_idx to be the calling thread
+ 
+void UpdatePreferredWorker(std::vector<int> &preferred_workers,
+                           unsigned par_idx) {
+  unsigned ran_on_idx = GetPerThread()->thread_id;
+  assert(ran_on_idx >= 0 && ran_on_idx < num_threads_);
+  assert(par_idx < preferred_workers.size());
+  preferred_workers[par_idx] = ran_on_idx;
+}
+
+// Schedule [par_idx_start,par_idx_end) across the preferred workers
+
+void ScheduleOnPreferredWorkers(PerThread& pt,
+                                ThreadPoolParallelSection& ps,
+                                std::vector<int> &preferred_workers,
+                                unsigned par_idx_start,
+                                unsigned par_idx_end,
+                                std::function<void(unsigned)> worker_fn) {
+  for (auto par_idx = par_idx_start; par_idx < par_idx_end; ++par_idx) {
+    // Look up hint for par_idx.  Note that the hints may have been
+    // recorded from a prior thread pool with a different number of
+    // threads, hence we must cap at num_threads_.
+    assert(par_idx < preferred_workers.size());
+    unsigned q_idx = preferred_workers[par_idx] % num_threads_;
+    assert(q_idx < num_threads_);
+    WorkerData& td = worker_data_[q_idx];
+    Queue& q = td.queue;
+    unsigned w_idx;
+
+    // Attempt to enqueue the task
+    auto pr = q.PushBackWithTag([worker_fn, par_idx, &preferred_workers, &ps, this]() {
+        UpdatePreferredWorker(preferred_workers, par_idx);
```

Comment:
>UpdatePreferredWorker [](http://example.com/codeflow?start=8&length=21)

Why we need to update it here? Are there any subsequent logics consumes preferred_workers[par_idx]?

Reply:
This is actually the most important piece, and I will add a comment to identify it :).  This is where the task records which worker thread executed it -- hence this is the _new_ preferred worker for use in the next loop.